### PR TITLE
Invoice status will be updated based on stripe response

### PR DESCRIPTION
**Pull request by gowsik-ragunath**
## Notion card

https://www.notion.so/saeloun/Fix-the-critical-issue-on-stripe-payments-267d5635c20844ab8ef4c4184021527c

## Summary

This PR adds a validation that will check if the payment for an invoice is recorded in Stripe, if the payment is recorded then invoice status will be updated to `Paid` if not, then an error message will be displayed.

## Preview

If the payment is recorded in Stripe:

Screenshot 2022-07-06 at 6 52 30 PM

If payment is not recorded in Stripe or user manually tries to visit payment success page:

Screenshot 2022-07-06 at 6 53 19 PM

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### Checklist:

- [X] I have manually tested all workflows
- [X] I have performed a self-review of my own code
- [X] I have added automated tests for my code

### CODE REVIEW

**Review by apoorv1316 — APPROVED** [review]
LGTM 🚀

**Comment by supriya3105** [comment]
@akhilgkrishnan Please review this PR.

File: app/models/invoice.rb [hunk 1]
```diff
@@ -2,22 +2,23 @@
...
+#  amount_paid           :decimal(20, 2)   default(0.0)
+#  discount              :decimal(20, 2)   default(0.0)
+#  due_date              :date
+#  external_view_key     :string
+#  invoice_number        :string
+#  issue_date            :date
+#  outstanding_amount    :decimal(20, 2)   default(0.0)
+#  reference             :text
+#  status                :integer          default("draft"), not null
+#  stripe_payment_intent :string
```
[the remark this answers is not in the document]
Comment (gowsik-ragunath) [reply app/models/invoice.rb]: @akhilgkrishnan As discussed in the call, I have removed `stripe_payment_intent` and replaced it with `payment info` jsonb field (store_accessor).

**Comment by supriya3105** [comment]
@akhilgkrishnan Please review the PR again and approve if it looks good.

**Review by akhilgkrishnan — APPROVED** [review]
LGTM

File: db/schema.rb [hunk 2]
```diff
@@ -129,6 +132,23 @@
...
+    t.datetime "expired_at"
+    t.string "first_name"
+    t.string "last_name"
+    t.integer "role", default: 0, null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["company_id"], name: "index_invitations_on_company_id"
+    t.index ["sender_id"], name: "index_invitations_on_sender_id"
+    t.index ["token"], name: "index_invitations_on_token", unique: true
+  end
```
Comment (keshavbiswa) [inline db/schema.rb:150]: This is not part of this PR.

**Review by keshavbiswa — CHANGES_REQUESTED** [review]

File: app/models/invoice.rb [hunk 3]
```diff
@@ -2,23 +2,23 @@
...
+#  amount             :decimal(20, 2)   default(0.0)
+#  amount_due         :decimal(20, 2)   default(0.0)
+#  amount_paid        :decimal(20, 2)   default(0.0)
+#  discount           :decimal(20, 2)   default(0.0)
+#  due_date           :date
+#  external_view_key  :string
+#  invoice_number     :string
+#  issue_date         :date
+#  outstanding_amount :decimal(20, 2)   default(0.0)
+#  payment_infos      :jsonb
```
[the remark this answers is not in the document]
Comment (gowsik-ragunath) [reply app/models/invoice.rb:15]: Akhil suggested that we can use this JSON field to store checkout payment reference data from different payment processors like stripe, wise, and Razorpay. Can we rename this field name to `checkout_info`, `payment_details` or `payment_references`?

[the remark this answers is not in the document]
Comment (keshavbiswa) [reply app/models/invoice.rb:15]: Ideally all invoice should have an associated payment/order record. Storing these important data in a jsonb column doesn't sound quite right to me.

[the remark this answers is not in the document]
Comment (gowsik-ragunath) [reply app/models/invoice.rb:15]: Yeah, but this payment record will be created only after confirming that the payment is successful but we need to store these checkout id, and payment intent id while creating a Stripe checkout session, once the payment is successful and users are redirected to our site we will use these ids to check if payment is successful or not by making API calls to Stripe.

File: db/schema.rb (hunk 2 above)
Comment (gowsik-ragunath) [reply db/schema.rb]: Removed invitations related database schemas.

I used `rails db:schema:load` while switching between the invitation branch and this branch to get the old devise inevitable column back in DB, not sure if this invitation data schema was added into schema.rb file while running this command.

File: app/models/invoice.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (gowsik-ragunath) [reply app/models/invoice.rb:15]: @keshavbiswa soft ping.

**Comment by supriya3105** [comment]
@keshavbiswa Reminder to review this PR.

**Review by keshavbiswa — APPROVED** [review]

**Pull request merged**